src/resolvelib/structs.py

```diff
@@ -62,3 +67,32 @@ def iter_children(self, key):
 
     def iter_parents(self, key):
         return iter(self._backwards[key])
+
+
+class Candidates(object):
+    """Wrap the candidate iterator factory returned by ``find_matches()``.
+
+    Calling `iter()` on this class would invoke the underlying iterator
+    factory, making it a "collection with ordering" that can be iterated
+    through multiple times, but lacks random access methods presented in
+    built-in Python sequence types.
+    """
+
+    def __init__(self, matches):
+        if callable(matches):
+            self._factory = matches
+        elif not isinstance(matches, collections_abc.Sequence):
+            self._factory = functools.partial(iter, list(matches))
+        else:
+            self._factory = functools.partial(iter, matches)
+
```

Comment:
`candidates` passed to `get_preference` isn’t a `Candidates` instance, but the underlying iterator, since I want to keep the class an implementation detail if possible. I also want to avoid a custom `__len__` since it be a performance trap for implementers when they provide an iterator.

OTOH, lazy evaluation isn’t always needed (in fact I’d argue Python package’s need of it is a niche case). All of the tests certainly don’t need it. So maybe it’d be better if we are able to pass the return value of `find_matches` transparently to `get_preference`, i.e. you get a list in `get_preference` if you provided the resolver a list, and an iterator if you provided an iterator. I’ll look into the possibility.

Reply:
> candidates passed to get_preference isn’t a Candidates instance

Ah, sorry. I missed that...